be/src/vec/functions/array/function_array_element.h

```diff
@@ -0,0 +1,247 @@
+        const auto& src_str_chars = nested_column->get_chars();
+
+        // prepare return data
+        auto dst = ColumnString::create();
+        auto& dst_str_offs = dst->get_offsets();
+        dst_str_offs.resize(input_rows_count);
+        auto& dst_str_chars = dst->get_chars();
+        dst_str_chars.reserve(src_str_chars.size());
+
+        // process
+        for (size_t row = 0; row < input_rows_count; ++row) {
+            size_t off = offsets[row - 1];
+            size_t len = offsets[row] - off;
+            auto index = arguments[1].column->get_int(row);
+            // array is nullable
+            bool null_flag = bool(src_null_map && src_null_map[row]);
+            // calc index in nested column
+            if (!null_flag && index > 0 && index <= len) {
+                index += off - 1;
+            } else if (!null_flag && index < 0 && -index <= len) {
+                index += off + len;
+            } else {
+                null_flag = true;
+            }
+            // nested column nullable check
+            if (!null_flag && nested_null_map && nested_null_map[index]) {
+                null_flag = true;
+            }
+            // actual string copy
+            if (!null_flag) {
+                DCHECK(index >= 0 && index < src_str_offs.size());
+                dst_null_map[row] = false;
+                auto element_size = src_str_offs[index] - src_str_offs[index - 1];
+                dst_str_offs[row] = dst_str_offs[row - 1] + element_size;
+                auto src_string_pos = src_str_offs[index - 1];
+                auto dst_string_pos = dst_str_offs[row - 1];
+                dst_str_chars.resize(dst_string_pos + element_size);
+                memcpy(&dst_str_chars[dst_string_pos], &src_str_chars[src_string_pos],
+                       element_size);
+            } else {
+                dst_null_map[row] = true;
+                dst_str_offs[row] = dst_str_offs[row - 1];
+            }
+        }
+
+        VLOG_DEBUG << "function:" << get_name() << ", array type:" << arguments[0].type->get_name()
+                   << ", index type:" << arguments[1].type->get_name()
+                   << ", result type:" << result_type->get_name();
+        return dst;
+    }
+
+    template <typename ElementDataType>
+    ColumnPtr _execute_number(const ColumnsWithTypeAndName& arguments,
+                              const DataTypePtr& result_type, size_t input_rows_count,
+                              const UInt8* src_null_map, UInt8* dst_null_map) {
+        // check array nested column type and get data
+        auto array_column = check_and_get_column<ColumnArray>(*arguments[0].column);
+        DCHECK(array_column != nullptr);
+        const ColumnVector<ElementDataType>* nested_column = nullptr;
+        const UInt8* nested_null_map = nullptr;
+        auto nested_null_column = check_and_get_column<ColumnNullable>(array_column->get_data());
```

Comment:
Both argument column and column inside array column can be nullable? It means it's possible for this function to pass argument like this `ColumnNullable(ColumnArray(ColumnNullable(ColumnInt8)))` ? I think it's better to de-duplicate these two ColumnNullable

Reply:
Yes we support ColumnNullable(ColumnArray(ColumnNullable(ColumnInt8)))

Reply:
@cambyzju I'm still confused about this. If outer ColumnNullable(null=true) is enough to indicate ColumnArray is null, it seems like the inner  ColumnNullable is redundant. So could you give an example which outer ColumnNullable is not null and inner ColumnNullable is null? Thx

Reply:
ColumnNullable(  // the array may be null or not
ColumnArray(
ColumnNullable(  // nested element may be null or not
ColumnInt8)))

for example:
```
> select * from array_test;
+------+------+--------+
| k1   | k2   | k3     |
+------+------+--------+
|    3 | NULL | NULL   |
|    1 |    2 | [1, 2, NULL] |
|    4 | NULL | []     |
+------+------+--------+
```
row 1 : array is null (outer ColumnNullable)
row 2: one element of the array is null (inner ColumnNullable)